anesthetic/read/montepythonreader.py

```diff
@@ -0,0 +1,158 @@
+"""Tools for reading from MontePython chains files."""
+import os
+import glob
+import warnings
+import numpy
+from anesthetic.read.chainreader import ChainReader
+try:
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        from montepython import analyze
+except ImportError:
+    pass
+
+
+class MontePythonReader(ChainReader):
+    """Read and process MontePython chain files using `montepython.analyze`."""
+
+    def __init__(self, root):
+        """Initialise MontePythonReader.
+
+        Note how MontePython takes the _folder_ to the chain files as root,
+        different from the root for GetDist.
+
+        Parameters
+        ----------
+        root: str
+            Path to _folder_ containing the MontePython .txt chain files.
+            E.g. for the following two chain files:
+            /path/to/chains/Planck2015_TTlowP/2019-04-29_100000__1.txt
+                                              2019-04-29_100000__2.txt
+            you should pass the string "/path/to/chains/Planck2015_TTlowP".
+            No trailing '/'.
+
+        """
+        super(MontePythonReader, self).__init__(root)
+
+    @property
+    def log_param_file(self):
+        """File containing input parameter info."""
+        return self.root + '/log.param'
+
+    @property
+    def paramnames_file(self):
+        """File containing parameter names."""
+        return glob.glob(self.root + '/*_.paramnames')[-1]
+
+    @property
+    def log_file(self):
+        """File containing MontePython log data."""
+        return self.root + '/' + os.path.basename(self.root) + '.log'
+
+    def _init(self):
+        """Prepare MontePython data.
+
+        Uses MontePython's `analyze` module to prepare the data:
+            * Extracts parameter names from MontePython's log.param file.
+            * Removes burn-in and non-markovian points from the data.
+            * Extract tex names.
+            * Extracts parameter limits from MontePython's log.param file.
+
+        """
+        # The variable and function names used here correspond to the ones
+        # used in MontePython's analyze.py module.
+        command_line = Namespace(files=[self.root])
+        self.info = analyze.Information(command_line)
+        analyze.prepare(files=command_line.files, info=self.info)
+        analyze.extract_parameter_names(info=self.info)
+        analyze.find_maximum_of_likelihood(info=self.info)
+        data_per_chain = analyze.remove_bad_points(info=self.info)
+        self.data = numpy.concatenate(data_per_chain, axis=0)
```

Comment:
Is there a reason why this code is not in the ``__init__`` function?

Reply:
Yes that was changed in commit [080ab4a ](https://github.com/williamjameshandley/anesthetic/pull/14/commits/080ab4a0d98748a95335b1fae821d53ba4d178f8):

> * Deferred the proper calls to `montepython.analyze` to the `_init`
>   method in `MontePythonReader` (previously already done in `__init__`)
>   such that an instance can be created and its properties used even if
>   the provided root actually is not a MontePython root. This allows the
>   necessary checks in `SampleReader` to be done.

This allows `SampleReader` to use `MontePythonReader`'s properties such as `log_param_file`, `paramnames_file` and `log_file` in order to determine whether the given root belongs to a MontePython folder.